### PR TITLE
[107049] Recognise top level keys in config.toml.example

### LINKED ISSUE #107049: `./configure --set changelog-seen=2` doesn't work
```
; ./configure --set changelog-seen=2
configure: processing command line
configure: 
configure: changelog-seen       := 2
configure: build.configure-args := ['--set', 'changelog-seen=2']
Traceback (most recent call last):
  File "/Users/jyn/src/rust/./src/bootstrap/configure.py", line 465, in <module>
    raise RuntimeError("config key {} not in sections".format(section_key))
RuntimeError: config key changelog-seen not in sections
```

The problem is that the configure script assumes that all keys are in a section and there are no top-level keys. This forces people to use `printf` instead: https://github.com/rust-lang/rust/blob/37c1d6dc09e97a7d707977c7cceeb0685df1776a/README.md#L89-L96

We should make it possible to set top-level keys through `configure`, both `changelog-seen` and `profile`. We should also update the line in the readme linked above to suggest `./configure` instead of printf.

### PR DESCRIPTION
**Pull request by lionellloh**
Closes #107049

Test Plan

Configure changelog-seen
```
lionellloh@lionellloh-mbp rust % ./configure --set changelog-seen=1
configure: processing command line
configure:
configure: changelog-seen       := 1
configure: build.configure-args := ['--set', 'changelog-seen=1']
configure:
configure: writing `config.toml` in current directory
configure:
configure: run `python /Users/lionellloh/rust/x.py --help`
lionellloh@lionellloh-mbp rust % diff config.toml config.toml.example
16c16
< changelog-seen = 1
---
> changelog-seen = 2
331c331
< configure-args = ['--set', 'changelog-seen=1']
---
> #configure-args = []
675c675
< [target.x86_64-apple-darwin]
---
> [target.x86_64-unknown-linux-gnu]
809d808
<
```

Configure profile

```
lionellloh@lionellloh-mbp rust % ./configure --set profile=xyz
configure: processing command line
configure:
configure: profile              := xyz
configure: build.configure-args := ['--set', 'profile=xyz']
configure:
configure: writing `config.toml` in current directory
configure:
configure: run `python /Users/lionellloh/rust/x.py --help`
lionellloh@lionellloh-mbp rust % diff config.toml config.toml.example
26c26
< profile = xyz
---
> #profile = <none>
331c331
< configure-args = ['--set', 'profile=xyz']
---
> #configure-args = []
675c675
< [target.x86_64-apple-darwin]
---
> [target.x86_64-unknown-linux-gnu]
809d808
<
```

### CODE REVIEW

File: src/bootstrap/configure.py [hunk 1]
```diff
@@ -459,12 +465,23 @@ def configure_section(lines, config):
                 raise RuntimeError("failed to find config line for {}".format(key))
 
 
-for section_key in config:
-    section_config = config[section_key]
```
Comment (lionellloh) [inline src/bootstrap/configure.py:463]: removed this and used `<dict>.items()` instead.

File: src/bootstrap/configure.py [hunk 2]
```diff
@@ -459,12 +465,23 @@ def configure_section(lines, config):
...
-    if section_key not in sections:
-        raise RuntimeError("config key {} not in sections".format(section_key))
+def configure_top_level_key(lines, top_level_key, value): 
+    for i, line in enumerate(lines): 
+        if line.startswith('#' + top_level_key + ' = ') or line.startswith(top_level_key + ' = '):
```
Comment (lionellloh) [inline src/bootstrap/configure.py:470]: account for commented and uncommented top-level key-values

File: src/bootstrap/configure.py [hunk 3]
```diff
@@ -459,12 +465,23 @@ def configure_section(lines, config):
                 raise RuntimeError("failed to find config line for {}".format(key))
 
 
-for section_key in config:
-    section_config = config[section_key]
-    if section_key not in sections:
-        raise RuntimeError("config key {} not in sections".format(section_key))
+def configure_top_level_key(lines, top_level_key, value):
```
Comment (lionellloh) [inline src/bootstrap/configure.py:468]: decided not to overload `configure_section()`

**Comment by lionellloh** [comment]
Noob here, any lints I can run to autoformat the configure.py code? Or do I have to fix it manually? @jyn514 / @zephaniahong

**Comment by jyn514** [comment]
@lionellloh you could run `black` manually, but there's no formatting enforced by CI, no

**Comment by lionellloh** [comment]
> @lionellloh you could run `black` manually, but there's no formatting enforced by CI, no

I am seeing this in the CI: 

`tidy error: /checkout/src/bootstrap/configure.py:481: trailing whitespace` 

Is that just a warning?

**Comment by zephaniahong** [comment]
@lionellloh There a tool called tidy that enforces some formatting: https://rustc-dev-guide.rust-lang.org/tests/intro.html?highlight=tidy#tidy
To check locally, run `./x.py test tidy`.

**Comment by zephaniahong** [comment]
@jyn514 Are you okay with running black? Because the last time I made a PR for a python file (and was formatted by some formatter), I was told to undo the formatting.

**Comment by lionellloh** [comment]
> @lionellloh There a tool called tidy that enforces some formatting: https://rustc-dev-guide.rust-lang.org/tests/intro.html?highlight=tidy#tidy To check locally, run `./x.py test tidy`.

Awesome, that's what I am looking for!

**Comment by lionellloh** [comment]
> > @lionellloh There a tool called tidy that enforces some formatting: https://rustc-dev-guide.rust-lang.org/tests/intro.html?highlight=tidy#tidy To check locally, run `./x.py test tidy`.
> 
> Awesome, that's what I am looking for!

```
FileNotFoundError: [Errno 2] No such file or directory: '/Users/lionellloh/rust/src/bootstrap/defaults/config.xyz.toml'
``` 

Is this expected?

**Comment by zephaniahong** [comment]
That's probably because of what you did when testing out your `./configure
Change your `config.toml` to this 
```toml
profile = "compiler"
changelog-seen = 2
```

**Comment by zephaniahong** [comment]
@lionellloh Once your tests pass, do squash your commits.

**Comment by lionellloh** [comment]
> @lionellloh Once your tests pass, do squash your commits.

Sure! Any comments on the PR?

**Comment by zephaniahong** [comment]
I'll leave that to the reviewer since I'm still relatively new to this. Just wanted to help you reach a point where it'd be easier for your reviewer!(:

**Comment by zephaniahong** [comment]
Also, if you're interested, you could consider working on #107050 as its closely related to this issue. You might want to check if the current assignee is still interested in working on it though.

**Comment by zephaniahong** [comment]
I don't think Jyn is free to review PRs. Usually, give the reviewer about 2 weeks. If you don't get a response, then do ping your reviewer. Alternatively, you can join the Zulip channel as sometimes reviewers are more active there.

**Comment by lionellloh** [comment]
Thanks everyone! I am keen to contribute more. Feel free to assign more issues to me. I will look at https://github.com/rust-lang/rust/issues/107050 next

**Comment by lionellloh** [comment]
Hey sorry yeah do assign it to me. I can take a look and fix it.

**Pull request merged**